File: python/harness/redis_harness.py
"""The Redis harness (RedisHarness): membership cache, unread counters, breaker
state, presence keys. Seed = set keys directly (pre-warm a cache to prove
invalidation). Assert = read keys/TTLs. Reset = FLUSHDB (the trivially fast
reset — contrast with Postgres TRUNCATE).
"""

from __future__ import annotations

import redis
from testcontainers.redis import RedisContainer

from harness.images import REDIS_IMAGE


class RedisHarness:
    def __init__(self) -> None:
        self._container: RedisContainer | None = None
        self._client: redis.Redis | None = None
        self._host = ""
        self._port = 0

    def start(self) -> None:
        self._container = RedisContainer(image=REDIS_IMAGE)
        self._container.start()
        self._host = self._container.get_container_host_ip()
        self._port = int(self._container.get_exposed_port(6379))
        self._client = redis.Redis(host=self._host, port=self._port, decode_responses=True)

    @property
    def client(self) -> redis.Redis:
        assert self._client is not None
        return self._client

    @property
    def host(self) -> str:
        return self._host

    @property
    def port(self) -> int:
        return self._port

    def members_cache_present(self, channel_id: str) -> bool:
        return self.client.exists("members:" + channel_id) > 0

    def cached_members(self, channel_id: str) -> set[str]:
        return set(self.client.smembers("members:" + channel_id))

    def reset(self) -> None:
        self.client.flushdb()

    def stop(self) -> None:
        if self._client is not None:
            self._client.close()
        if self._container is not None:
            self._container.stop()

File: typescript/harness/redis.harness.ts
// The Redis harness (cache / counters / breaker state). Real container.
// Seed = set keys directly (e.g. pre-warm a membership cache to prove
// invalidation); Assert = read keys/TTLs; Reset = FLUSHDB (the trivially fast
// reset — contrast with Postgres's TRUNCATE).

import { RedisContainer, type StartedRedisContainer } from '@testcontainers/redis';
import Redis from 'ioredis';

import type { DependencyHarness } from './dependency-harness.js';
import { REDIS_IMAGE } from './images.js';

export class RedisHarness implements DependencyHarness {
  private container?: StartedRedisContainer;
  private redisClient?: Redis;
  private host = '';
  private port = 0;

  get client(): Redis {
    if (!this.redisClient) {
      throw new Error('RedisHarness not started');
    }
    return this.redisClient;
  }

  get address(): { host: string; port: number } {
    return { host: this.host, port: this.port };
  }

  async start(): Promise<void> {
    this.container = await new RedisContainer(REDIS_IMAGE).start();
    this.host = this.container.getHost();
    this.port = this.container.getMappedPort(6379);
    this.redisClient = new Redis({ host: this.host, port: this.port });
  }

  async reset(): Promise<void> {
    await this.client.flushdb();
  }

  async stop(): Promise<void> {
    if (this.redisClient) {
      this.redisClient.disconnect();
    }
    if (this.container) {
      await this.container.stop();
    }
  }

  // seedMembershipCache pre-warms members:{channelId} so a test can prove
  // invalidation (G-CACHE).
  async seedMembershipCache(channelId: string, ...memberIds: string[]): Promise<void> {
    const key = `members:${channelId}`;
    const pipe = this.client.multi();
    pipe.del(key);
    if (memberIds.length > 0) {
      pipe.sadd(key, ...memberIds);
      pipe.expire(key, 300);
    }
    await pipe.exec();
  }

  // cacheHasMember reports whether the cached set exists and contains a user.
  async cacheHasMember(channelId: string, userId: string): Promise<{ exists: boolean; member: boolean }> {
    const key = `members:${channelId}`;
    const exists = await this.client.exists(key);
    if (exists === 0) {
      return { exists: false, member: false };
    }
    const member = await this.client.sismember(key, userId);
    return { exists: true, member: member === 1 };
  }
}
